druid/src/widget/immediate.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{widget::prelude::*, Data};
+
+/// A widget that allows writing parts of the ui in an immediate-mode style.
+///
+/// The primary usecase for this is displaying data that doesn't fit into druids model, such as
+/// enums you have no control over.
+///
+/// Whenever the state represented by the `Immediate` changes,
+/// it will reconstruct its content for the new data.
+///
+/// While this is not the most efficient thing to do, it is very simple and performs perfectly fine
+/// for small or rarely changed data.
+///
+/// Note that this is more like a last resort when the data to be represented is in a for druid particularly
+/// unfitting format.
+pub struct Immediate<D, W: Widget<()>> {
+    constructor: Box<dyn Fn(&D) -> W>,
+    content: Option<W>,
+}
+
+impl<D, W: Widget<()>> Immediate<D, W> {
+    /// Takes a constructor for a stateless widget
+    pub fn new(constructor: impl Fn(&D) -> W + 'static) -> Self {
+        Self {
+            constructor: Box::new(constructor),
+            content: None,
+        }
+    }
+}
+
+impl<D: Data, W: Widget<()>> Widget<D> for Immediate<D, W> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, _data: &mut D, env: &Env) {
+        if let Some(content) = &mut self.content {
+            content.event(ctx, event, &mut (), env);
+        }
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &D, env: &Env) {
+        if let LifeCycle::WidgetAdded = event {
+            self.content = Some((self.constructor)(data));
+        }
+        if let Some(content) = &mut self.content {
+            content.lifecycle(ctx, event, &(), env);
+        }
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &D, data: &D, env: &Env) {
+        if !old_data.same(data) {
+            self.content = Some((self.constructor)(data));
```

Comment:
A `LifeCycle::WidgetAdded` event should be sent to the newly constructed widget.

Reply:
Doesn't this happen when I call `ctx.children_changed()`?

Reply:
I think `ctx.children_changed()` will trigger a `LifeCycle::RouteWidgetAdded` but not `LifeCycle::WidgetAdded`. Having a `WidgetPod` encapsulate `self.content` might also fix it but seems heavyweight for `Immediate`.

I haven't researched the whole `WidgetAdded` system too much, so perhaps @cmyr can chime in here. However I think that we could even remove `ctx.children_changed()` and just send the `WidgetAdded` event after creating the inner widget.

Reply:
yes, all you need to do after adding a widget is call `ctx.children_changed()`. `RouteWidgetAdded` is an implementation detail; when we see that event, we check to see if the current widget is new, and if it is we replace it with `WidgetAdded`.

Reply:
So this would work as expected? Would there be any benefit in sending the `WidgetAdded` event directly?

Reply:
This is correct. The only time it would be okay to send `WidgetAdded` directly is when we are already handling a `WidgetAdded` event; otherwise state will not correctly propagate back up the tree.

Reply:
I looked more closely at druid code and my initial comment seems correct. The only place in the whole druid code base that generates `WidgetAdded` is the `WidgetPod` if it is new.

Thus saying that *all you need to do after adding a widget is call `ctx.children_changed()`* is not true. Your widgets need to also be encapsulated in a `WidgetPod` for it to work.

So let's clarify, @cmyr. Are you saying that `WidgetPod` use is mandatory and bare widgets are no longer valid druid code?

Reply:
sorry, this stuff really needs to be written down.

Basically: for a widget to fully participate in druid (to have an id, to have its own `hot` state, to be focusable, to correctly handle events) that widget needs to be in a `WidgetPod`. One way of thinking about `WidgetPod` is that it is the 'base class' for a widget.

In general, `WidgetPod`s are created behind the scenes; for instance a container widget creates a `WidgetPod` for each widget it manages.

You *can*, in certain circumstances, use a widget on its own. The main example of this I can think of is when a widget wants to reuse another widgets `paint` method, such as how `Button` currently owns a `Label` that is not in a pod; the button calls methods on the label directly, paints it with an offset, etc.

So yes, looking at this, if we want these widgets to correctly receive `WidgetAdded` we will need each one to be wrapped in a widgetpod.

Reply:
Thanks for the description. It matches my understanding.

Now back to what I was originally saying. One solution is to have `self.content` be encapsulated in a `WidgetPod`. It's certainly valid and also the most robust solution. However because we're really just passing events straight to `self.content` with no processing we could skip having a `WidgetPod` for performance reasons. That means `Immediate` would have to emulate the `WidgetAdded` lifecycle event for `self.content` whenever it gets re-created. It's really straightforward to do and won't cause any issues. The only threat of issues is when `WidgetPod` gets more bootstrapping functionality like `WidgetAdded` and we forget to update `Immediate`.

Reply:
Yes, this would be possible. If we wanted our child to do anything at all fancy, though (such as receive events or set timers or even potentially animate) then we struggle to make that work.

Reply:
Why would we struggle? If `Immediate` can do it, then the child can do it. Because we just lend out the `Immediate` identity/ctx to the child and `Immediate` itself never does anything.

Reply:
For one thing, If the widget was a container, its children would not be appropriately registered, and so things like focus would break.